src/core/meshcop/tcat_agent.cpp

```diff
@@ -629,13 +656,135 @@ Error TcatAgent::HandleGetProvisioningUrl(Message &aOutgoingMessage, bool &respo
+    pskc = datasetInfo.Get<Dataset::kPskc>();
+
+    error = VerifyHash(aIncomingMessage, aOffset, aLength, reinterpret_cast<const char *>(pskc.m8), Pskc::kSize);
+
+exit:
+    return error;
+}
+
+Error TcatAgent::HandlePresentInstallCodeHash(const Message &aIncomingMessage, uint16_t aOffset, uint16_t aLength)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(mVendorInfo->mInstallCode != nullptr, error = kErrorSecurity);
+
+    error = VerifyHash(aIncomingMessage, aOffset, aLength, mVendorInfo->mInstallCode,
+                       StringLength(mVendorInfo->mInstallCode, 255));
+
+exit:
+    return error;
+}
+
+Error TcatAgent::HandleRequestRandomNumberChallenge(Message &aOutgoingMessage, bool &response)
+{
+    Error error = kErrorNone;
+
+    SuccessOrExit(error = Random::Crypto::Fill(mRandomChallenge));
+
+    SuccessOrExit(
+        error = Tlv::AppendTlv(aOutgoingMessage, kTlvResponseWithPayload, &mRandomChallenge, sizeof(mRandomChallenge)));
+    response = true;
+exit:
+    return error;
+}
+
+Error TcatAgent::HandleRequestPskdHash(const Message &aIncomingMessage,
+                                       Message       &aOutgoingMessage,
+                                       uint16_t       aOffset,
+                                       uint16_t       aLength,
+                                       bool          &response)
+{
+    Error                    error             = kErrorNone;
+    uint64_t                 providedChallenge = 0;
+    Crypto::HmacSha256::Hash hash;
+
+    VerifyOrExit(StringLength(mVendorInfo->mPskdString, kMaxPSKdLength) != 0, error = kErrorFailed);
+    VerifyOrExit(aLength == sizeof(providedChallenge), error = kErrorParse);
+
+    SuccessOrExit(error = aIncomingMessage.Read(aOffset, &providedChallenge, aLength));
+
+    hash = CalculateHash(providedChallenge, mVendorInfo->mPskdString,
+                         StringLength(mVendorInfo->mPskdString, kMaxPSKdLength));
+
+    SuccessOrExit(error = Tlv::AppendTlv(aOutgoingMessage, kTlvResponseWithPayload, hash.GetBytes(),
+                                         Crypto::HmacSha256::Hash::kSize));
     response = true;
 
 exit:
     return error;
 }
 
+Error TcatAgent::VerifyHash(const Message &aIncomingMessage,
+                            uint16_t       aOffset,
+                            uint16_t       aLength,
+                            const char    *aBuf,
+                            size_t         aBufLen)
+{
+    Error                    error = kErrorNone;
+    Crypto::HmacSha256::Hash hash;
+    Crypto::HmacSha256::Hash recivedHash;
+
+    VerifyOrExit(aLength == Crypto::HmacSha256::Hash::kSize, error = kErrorSecurity);
+    VerifyOrExit(mRandomChallenge != 0, error = kErrorSecurity);
+
+    SuccessOrExit(error = aIncomingMessage.Read(aOffset, recivedHash.m8, aLength));
+
+    hash = CalculateHash(mRandomChallenge, aBuf, aBufLen);
+
+    VerifyOrExit(hash == recivedHash, error = kErrorSecurity);
```

Comment:
This works fine, but we can also directly compare the calcualted `hash` with the bytes in the message without the need to read it from `aIncomingMessage`?

We have `Message::Compare<ObjectType>()` and `Message::CompareBytes()`. For example:

```c++
    VerifyOrExit(aIncomingMessage.Compare(aOffset, hash), error = kErrorSecurity);
```

Reply:
fixed.